Add some conversion trait impls

Jules-Bertholet opened the pull request:
- `impl<T, const N: usize> From<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]>` (cc https://github.com/rust-lang/rust/issues/96097)
- `impl<T, const N: usize> AsRef<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]>` (cc https://github.com/rust-lang/rust/issues/96097)
- `impl<T, const N: usize> AsRef<[MaybeUninit<T>]> for MaybeUninit<[T; N]>` (cc https://github.com/rust-lang/rust/issues/96097)
- `impl<T, const N: usize> AsMut<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]>` (cc https://github.com/rust-lang/rust/issues/96097)
- `impl<T, const N: usize> AsMut<[MaybeUninit<T>]> for MaybeUninit<[T; N]>` (cc https://github.com/rust-lang/rust/issues/96097)
- `impl<T, const N: usize> From<MaybeUninit<[T; N]>> for [MaybeUninit<T>; N]` (cc https://github.com/rust-lang/rust/issues/96097)
- `impl<T, const N: usize> AsRef<[Cell<T>; N]> for Cell<[T; N]>` (equivalent of `as_array_of_cells`, cc https://github.com/rust-lang/rust/issues/88248)
- `impl<T, const N: usize> AsRef<[Cell<T>]> for Cell<[T; N]>`
- `impl<T> AsRef<[Cell<T>]> for Cell<[T]>` (equivalent of `as_slice_of_cells`)

@rustbot label T-libs-api needs-fcp -T-libs

I’ve tried to only add impls that are unlikely to cause single-applicable-impl inference breakage.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17010476775/job/48225855617) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48225855617) [(plain)](https://github.com/rust-lang/rust/commit/7ef29d9102f198137ee774da0259f2b2e3796c28/checks/48225855617/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17011001191/job/48227048621) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48227048621) [(plain)](https://github.com/rust-lang/rust/commit/3a9660420fee759f0d88f2d24cec0292af11bf79/checks/48227048621/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/suggestions/issue-71394-no-from-impl.rs stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/suggestions/issue-71394-no-from-impl/issue-71394-no-from-impl.stderr`
diff of stderr:

5    |                         ^^^^ the trait `From<&[u8]>` is not implemented for `&[i8]`
6    |
7    = help: the following other types implement trait `From<T>`:
+              `[MaybeUninit<T>; N]` implements `From<MaybeUninit<[T; N]>>`
8              `[T; 10]` implements `From<(T, T, T, T, T, T, T, T, T, T)>`
9              `[T; 11]` implements `From<(T, T, T, T, T, T, T, T, T, T, T)>`
10              `[T; 12]` implements `From<(T, T, T, T, T, T, T, T, T, T, T, T)>`

12              `[T; 2]` implements `From<(T, T)>`
13              `[T; 3]` implements `From<(T, T, T)>`
14              `[T; 4]` implements `From<(T, T, T, T)>`
-              `[T; 5]` implements `From<(T, T, T, T, T)>`
-            and 6 others
+            and 7 others
17    = note: required for `&[u8]` to implement `Into<&[i8]>`
18 
19 error: aborting due to 1 previous error


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args suggestions/issue-71394-no-from-impl.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/suggestions/issue-71394-no-from-impl.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/suggestions/issue-71394-no-from-impl" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0277]: the trait bound `&[i8]: From<&[u8]>` is not satisfied
##[error]  --> /checkout/tests/ui/suggestions/issue-71394-no-from-impl.rs:8:25
   |
LL |     let _: &[i8] = data.into();
   |                         ^^^^ the trait `From<&[u8]>` is not implemented for `&[i8]`
   |
   = help: the following other types implement trait `From<T>`:
             `[MaybeUninit<T>; N]` implements `From<MaybeUninit<[T; N]>>`
             `[T; 10]` implements `From<(T, T, T, T, T, T, T, T, T, T)>`
             `[T; 11]` implements `From<(T, T, T, T, T, T, T, T, T, T, T)>`
             `[T; 12]` implements `From<(T, T, T, T, T, T, T, T, T, T, T, T)>`
             `[T; 1]` implements `From<(T,)>`
             `[T; 2]` implements `From<(T, T)>`
             `[T; 3]` implements `From<(T, T, T)>`
             `[T; 4]` implements `From<(T, T, T, T)>`
           and 7 others
   = note: required for `&[u8]` to implement `Into<&[i8]>`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
```

</details>

jieyouxu commented:
r? libs-api

clarfonthey commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -1452,6 +1452,58 @@ impl<T, const N: usize> MaybeUninit<[T; N]> {
+    #[inline(always)]
+    fn from(arr: [MaybeUninit<T>; N]) -> Self {
+        // SAFETY: T and MaybeUninit<T> have the same layout
+        unsafe { intrinsics::transmute_unchecked(arr) }
```
Isn't this `MaybeUninit::transpose`?

clarfonthey commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -1452,6 +1452,58 @@ impl<T, const N: usize> MaybeUninit<[T; N]> {
+    #[inline(always)]
+    fn from(arr: MaybeUninit<[T; N]>) -> Self {
+        // SAFETY: T and MaybeUninit<T> have the same layout
+        unsafe { intrinsics::transmute_unchecked(arr) }
```
Ditto here, I believe this is also `transpose`.

bors commented:
:umbrella: The latest upstream changes (presumably #146494) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Jules-Bertholet commented:
> Could you elaborate on these, and point to where they might have been discussed? Was there a discussion of those specifically, or are you just proposing them because `Char::as_str` exists?

Yes.

> If so, given that `[Char]::as_str` _also_ exists, could you please provide the corresponding traits consistently for arrays/slices of Char?

Unfortunately, `alloc` fails to build (“type annotations needed”) when I do that.

clarfonthey commented:
I think that dedicated methods are nice because they make things a little clearer than just `into()`. We're allowed to have both.

tgross35 reviewed (changes requested):
As mentioned above, these could be `#[inline]` and not `#inline(always)]`. That and a squash then r=me

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/145504/52618eb338609df44978b0ca4451ab7941fd1c7a..30a32eb015b755738831352fe934ca6261ba0c58)*

tgross35 commented:
@Jules-Bertholet gentle ping for the small request above

Jules-Bertholet committed:
Add some conversion trait impls

- `impl<T, const N: usize> From<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]>`
- `impl<T, const N: usize> AsRef<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]>`
- `impl<T, const N: usize> AsRef<[MaybeUninit<T>]> for MaybeUninit<[T; N]>`
- `impl<T, const N: usize> AsMut<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]>`
- `impl<T, const N: usize> AsMut<[MaybeUninit<T>]> for MaybeUninit<[T; N]>`
- `impl<T, const N: usize> From<MaybeUninit<[T; N]>> for [MaybeUninit<T>; N]`
- `impl<T, const N: usize> AsRef<[Cell<T>; N]> for Cell<[T; N]>`
- `impl<T, const N: usize> AsRef<[Cell<T>]> for Cell<[T; N]>`
- `impl<T> AsRef<[Cell<T>]> for Cell<[T]>`

tgross35 reviewed (approved):
Thanks!

@bors r+ rollup

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/145504/c7f5f3e0d5defe632d44743cbaed56272e2b67f0..d0b3a0bdc02924cdd28e37f6209ddea0b6bd4727)*

nazar-pc commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -1532,6 +1532,56 @@ impl<T, const N: usize> MaybeUninit<[T; N]> {
+}
+
+#[stable(feature = "more_conversion_trait_impls", since = "CURRENT_RUSTC_VERSION")]
+impl<T, const N: usize> AsMut<[MaybeUninit<T>; N]> for MaybeUninit<[T; N]> {
```
Is there a strong reason to not have `impl<T, const N: usize> AsMut<MaybeUninit<[T; N]>> for [MaybeUninit<T>; N]` too?

And I have a few places where bridging APIs could benefit from `impl<T, const N: usize> AsMut<MaybeUninit<[T; N]>> for [T; N]` too, which today requires `transmute()`.
